src/linux_android.rs

```diff
@@ -1,7 +1,73 @@
 //! Implementation for Linux / Android without `/dev/urandom` fallback
-use crate::{util_libc, Error};
+use crate::{util_unix::sys_fill_exact, Error};
 use core::mem::MaybeUninit;
 
 pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    util_libc::sys_fill_exact(dest, util_libc::getrandom_syscall)
+    sys_fill_exact(dest, getrandom_syscall)
+}
+
+// The value of `EINTR` is not architecture-specific. It is checked against
+// `libc::EINTR` by linux_android_with_fallback.rs.
+pub const EINTR: i32 = 4;
+
+// Also used by linux_android_with_fallback to check if the syscall is available.
+cfg_if! {
+    // Assume Android always has libc available and always go through libc on
+    // Android to support any future possible restrictions on direct syscall
+    // access by the Android sandbox.
+    //
+    // TODO: Expand inilne assembly to other architectures to avoid depending
+    // on libc on Linux.
+    if #[cfg(all(target_os = "linux", target_arch = "x86_64"))] {
+        type Word = u64;
+        type IWord = i64;
+
+        #[allow(non_upper_case_globals)]
+        pub const SYS_getrandom: IWord = if cfg!(target_arch = "x86_64") { 318 } else { unimplemented!() };
+
+        pub fn getrandom_syscall(buf: &mut [MaybeUninit<u8>]) -> Result<usize, Error> {
+            const _:() = assert!(core::mem::size_of::<Word>() == core::mem::size_of::<usize>());
+
+            let mut ret: IWord;
+            let flags = 0;
+            unsafe {
+                core::arch::asm!(
+                    "syscall",
+                    in("rax") SYS_getrandom,
+                    in("rdi") buf.as_mut_ptr(),
+                    in("rsi") buf.len(),
+                    in("rdx") flags,
+                    lateout("rcx") _,
+                    lateout("r11") _,
+                    lateout("rax") ret,
+                    options(nostack),
```

Comment:
One thing to keep in mind when reviewing this: I don't know what I'm doing :)

Reply:
I used https://www.felixcloutier.com/x86/syscall to understand what syscall clobbers.